mlflow/store/sqlalchemy_store.py

```diff
@@ -248,102 +278,104 @@ def _check_run_is_deleted(self, run):
 
     def get_metric_history(self, run_uuid, metric_key):
-        metrics = self.session.query(SqlMetric).filter_by(run_uuid=run_uuid, key=metric_key).all()
-        return [metric.to_mlflow_entity() for metric in metrics]
+        with self.ManagedSessionMaker() as session:
+            metrics = session.query(SqlMetric).filter_by(run_uuid=run_uuid, key=metric_key).all()
+            return [metric.to_mlflow_entity() for metric in metrics]
 
     def log_param(self, run_uuid, param):
-        run = self._get_run(run_uuid)
-        self._check_run_is_active(run)
-        # if we try to update the value of an existing param this will fail
-        # because it will try to create it with same run_uuid, param key
-        try:
-            # This will check for various integrity checks for params table.
-            # ToDo: Consider prior checks for null, type, param name validations, ... etc.
-            self._get_or_create(SqlParam, run_uuid=run_uuid, key=param.key,
-                                value=param.value)
-        except sqlalchemy.exc.IntegrityError as ie:
-            # Querying metrics from run entails pushing the query down to DB layer.
-            # Hence the rollback.
-            self.session.rollback()
-            existing_params = [p.value for p in run.params if p.key == param.key]
-            if len(existing_params) == 0:
-                raise MlflowException("Log param request failed for run ID={}. Attempted to log"
-                                      " param={}. Error={}".format(run_uuid,
-                                                                   (param.key, param.value),
-                                                                   str(ie)))
-            else:
-                old_value = existing_params[0]
-                raise MlflowException("Changing param value is not allowed. Param with key='{}' was"
-                                      " already logged with value='{}' for run ID='{}. Attempted "
-                                      " logging new value '{}'.".format(param.key, old_value,
-                                                                        run_uuid, param.value),
-                                      INVALID_PARAMETER_VALUE)
+        with self.ManagedSessionMaker() as session:
+            run = self._get_run(run_uuid=run_uuid, session=session)
+            self._check_run_is_active(run)
+            # if we try to update the value of an existing param this will fail
+            # because it will try to create it with same run_uuid, param key
+            try:
+                # This will check for various integrity checks for params table.
+                # ToDo: Consider prior checks for null, type, param name validations, ... etc.
+                self._get_or_create(model=SqlParam, session=session, run_uuid=run_uuid,
+                                    key=param.key, value=param.value)
+            except sqlalchemy.exc.IntegrityError as ie:
```

Comment:
Same here -- does it make sense to use a new session?

Reply:
Discussed offline - it seems to make sense to use a single session and avoid attempting to transact in an invalid state by adding an explicit `commit()` statement before the `except` statement and an explicit `rollback()` after the `except` statement. I've made these changes and added substantial explanatory comments, as above.